### PR TITLE
Waiting for free saucelabs machines implemented (closes #23)

### LINKED ISSUE #23: Wait for free saucelabs machines before put the tests to the queue

### PR DESCRIPTION
**Pull request by georgiy-abbasov**
\cc @AlexanderMoskovkin 
PR

### CODE REVIEW

**Comment by georgiy-abbasov** [comment]
FPR

File: src/saucelabs/request.js [hunk 1]
```diff
@@ -39,4 +40,28 @@ export default class SaucelabsRequestAdapter {
 
         return body;
     }
+
+    async get (url) {
```
Comment (AlexanderMoskovkin) [inline src/saucelabs/request.js:48]: `get` and `put` methods contains duplicated code. Move it to the `async _request(params)` function

File: src/saucelabs/runner.js [hunk 2]
```diff
@@ -1,9 +1,10 @@
 import Promise from 'pinkie';
 import Job from './job';
 import wait from '../utils/wait';
+import SaucelabsRequestAdapter from './request';
 
-
-const REPORTING_TIMEOUT = 1000 * 30;
+const REPORTING_TIMEOUT       = 1000 * 30;
+const MACHINE_WAITING_TIMEOUT = 1000 * 60;
```
Comment (AlexanderMoskovkin) [inline src/saucelabs/runner.js]: `WAITING_FOR_FREE_MACHINE_TIMEOUT` will be better imho

File: src/saucelabs/runner.js [hunk 3]
```diff
@@ -53,31 +82,58 @@ export default class SauceLabsRunner {
...
+
+                    if (freeMachinesCount)
+                        this._startJobs(freeMachinesCount);
+                }
+
+                await wait(MACHINE_WAITING_TIMEOUT);
+            }
+        }
+    }
+
```
Comment (AlexanderMoskovkin) [inline src/saucelabs/runner.js:129]: It's all a little complicated. Also I can't get what the difference between `_startJobs` and `_runJobs` methods by their names. It sounds like the same thing. I propose the following structure:

``` js
export default class extends EventEmitter {
    constructor () {
        //...
    }

    _onJobDone (job, res) {
        //add result to this.jobResults
        this.emit('job-done');
    }

    _runNextJobs (count) {
        var jobs = getQueuedJobs(count); // get it from this.jobs by state

        if (jobs.length) {
            jobs.forEach(job => {
                job.run()
                    .then(res => this._onJobDone(job, res));
            });
        }
        else
            this._stopFreeMachinesChecker();
    }

    async _checkForFreeMachines () {
        var freeMachineCount = await this._getFreeMachineCount();

        if (freeMachineCount)
            this.emit('free-machines-explored', { count: freeMachineCount });
    }

    _startFreeMachinesChecker () {
        this._checkForFreeMachines();
        this.checkForFreeMachinesInterval = window.setInterval(()=> this._checkForFreeMachines(), INTERVAL);
    }

    _stopFreeMachinesChecker () {
        // ...
    }

    runTests () {
        return new Promise(resolve => {
            //...
            this._runJobs();

            this.on('free-machines-explored', (e) => this._runNextJobs(e.count));
            this.on('job-done', () => this._runNextJobs(1));
            this.on('done', () => resolve(this.jobResults));
        });
    }
}

```

**Comment by georgiy-abbasov** [comment]
FPR

File: src/saucelabs/runner.js [hunk 4]
```diff
@@ -53,31 +80,81 @@ export default class SauceLabsRunner {
...
+        if (freeMachinesCount)
+            this.emit('free-machines-found', { count: freeMachinesCount });
+    }
+
+    async _startNextJobs (count) {
+        var jobs = this._getQueuedJobs(count);
+
+        if (jobs.length) {
+            jobs.forEach(job => job
```
Comment (AlexanderMoskovkin) [inline src/saucelabs/runner.js]:
``` js
job
    .run()
    ....
```

is a multiline function, so we should wrap it with `{}`

``` js
jobs.forEach(job => {
    return job
        .run()
        ...
}
```

**Comment by georgiy-abbasov** [comment]
FPR

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Pull request merged**